File: java/pen/AutoPen.java
public class AutoPen extends Pen {
    private boolean isOpen;

    public AutoPen(int capacity) {
        super(capacity);
        this.isOpen = false;
    }

    public AutoPen() {
        this(4096);
    }

    public boolean open() {
        return this.isOpen;
    }

    public void click() {
        this.isOpen = !this.isOpen;
    }

    @Override
    public void write(Paper paper, String message) throws
        OutOfInkException, OutOfSpaceException, ClosedPenException
    {
        if ( !this.open() ) {
            throw new ClosedPenException();
        }
        super.write(paper, message);
    }
}

File: pen/AutoPen.cpp
#include "AutoPen.hpp"

AutoPen::AutoPen(int capacity, const char* title)
    : Pen(capacity, title), isOpen(false) {}

AutoPen::~AutoPen() {}

bool AutoPen::open() const {
    return this->isOpen;
}

void AutoPen::click() {
    this->isOpen = !this->isOpen;
}

void AutoPen::write(Paper& paper, const std::string& message) {
    if ( !this->open() ) {
        throw ClosedPenException();
    }
    Pen::write(paper, message);
}


std::ostream& operator<<(std::ostream& out, const AutoPen& ap) {
    std::cout << (Pen) ap;
    std::cout << (( ap.open() ) ? " open" : " closed");

    return out;
}